### PR TITLE
Registered/Unregistered User changes for create and join lobby

### PR DESCRIPTION
**Pull request by apple00juice**
Added unified LobbyDeltaWebSocketDTO for lobby
Changed name generation, to generate a name that is unique in the lobby
Added if name is already in the lobby, and we try to add a new player, change the player with the name conflict if he is not registered
Added check that unregistered users can't create a ranked lobby
Added check that unregistered users can't join a ranked lobby
Added test

### CODE REVIEW

File: src/main/java/ch/uzh/ifi/hase/soprafs22/exceptions/LobbyNameConflictException.java [hunk 1]
```diff
@@ -0,0 +1,4 @@
+package ch.uzh.ifi.hase.soprafs22.exceptions;
+
+public class LobbyNameConflictException extends Exception{
```
Comment (hilalcomak) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/exceptions/LobbyNameConflictException.java:3]: Shouldn't this exception hold the name that caused the conflict?

File: src/main/java/ch/uzh/ifi/hase/soprafs22/exceptions/LobbyRegisteredUsersNameConflictException.java [hunk 2]
```diff
@@ -0,0 +1,4 @@
+package ch.uzh.ifi.hase.soprafs22.exceptions;
+
+public class LobbyRegisteredUsersNameConflictException extends LobbyNameConflictException{
```
Comment (hilalcomak) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/exceptions/LobbyRegisteredUsersNameConflictException.java:3]: Shouldn't this exception hold the name that caused the conflict?

File: src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/Lobby.java [hunk 3]
```diff
@@ -250,12 +254,62 @@ public Player generatePlayer() {
         return new Player(generatedId, username, token, team);
     }
 
+    private String generateUniqueNameInLobby(){
+        String playerName = "Player-";
```
Comment (hilalcomak) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/Lobby.java:258]: Just an idea, have a list of names, a list of adjective and generate the name as randomAdjective + randomName. No need to do it.

File: src/main/java/ch/uzh/ifi/hase/soprafs22/rest/dto/web_socket/LobbyDeltaWebSocketDTO.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+import ch.uzh.ifi.hase.soprafs22.game.player.IPlayer;
+
+import java.util.List;
+
+/**
+ * Delta to be sent through web socket about the lobby so client can process info easier.
+ */
+public class LobbyDeltaWebSocketDTO {
+    private boolean pullUpdate;
```
Comment (hilalcomak) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/rest/dto/web_socket/LobbyDeltaWebSocketDTO.java:11]: I don't understand what this does. Is this requesting the client to fully pull the lobby?

File: src/main/java/ch/uzh/ifi/hase/soprafs22/service/LobbyService.java [hunk 5]
```diff
@@ -18,6 +19,8 @@
 import org.springframework.web.server.ResponseStatusException;
 
 import java.util.Collection;
+import java.util.LinkedList;
```
Comment (hilalcomak) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/service/LobbyService.java:22]: I think this is not used

File: src/test/java/ch/uzh/ifi/hase/soprafs22/lobby/LobbyTest.java [hunk 6]
```diff
@@ -98,5 +101,37 @@ void updateLobbyPlayer_userNameNotUnique_conflict() throws FullLobbyException {
...
+        lobby.addPlayer(player);
+
+        player = lobby.generatePlayer();
+        player.setName("Player-2");
+        System.out.println(player.getName());
+        lobby.addPlayer(player);
+
+        System.out.println("");
+
+        for (IPlayer p : lobby) {
```
Comment (hilalcomak) [inline src/test/java/ch/uzh/ifi/hase/soprafs22/lobby/LobbyTest.java:130]: I think you forgot the assert part of the test.

**Review by hilalcomak — APPROVED** [review]
Thank you very much. It looks awesome!

File: src/main/java/ch/uzh/ifi/hase/soprafs22/exceptions/LobbyNameConflictException.java (hunk 1 above)
Comment (apple00juice) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/exceptions/LobbyNameConflictException.java:3]: Sure, I'll add it

File: src/main/java/ch/uzh/ifi/hase/soprafs22/exceptions/LobbyRegisteredUsersNameConflictException.java (hunk 2 above)
Comment (apple00juice) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/exceptions/LobbyRegisteredUsersNameConflictException.java:3]: Sure, I'll add it

File: src/main/java/ch/uzh/ifi/hase/soprafs22/rest/dto/web_socket/LobbyDeltaWebSocketDTO.java (hunk 4 above)
Comment (apple00juice) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/rest/dto/web_socket/LobbyDeltaWebSocketDTO.java:11]: Yes, that some things changed and the client should fetch the lobby again. I was thinking about a good name, but that was the best I could come up with. Do you have any suggestions?

File: src/main/java/ch/uzh/ifi/hase/soprafs22/service/LobbyService.java (hunk 5 above)
Comment (apple00juice) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/service/LobbyService.java:22]: Will remove it

File: src/test/java/ch/uzh/ifi/hase/soprafs22/lobby/LobbyTest.java (hunk 6 above)
Comment (apple00juice) [reply src/test/java/ch/uzh/ifi/hase/soprafs22/lobby/LobbyTest.java:130]: Ahh, that was just for testing. I will remove that

Comment (apple00juice) [reply src/test/java/ch/uzh/ifi/hase/soprafs22/lobby/LobbyTest.java:130]: I quickly added two tests for testing unique name generation and resolving name conflict

File: src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/PlayerJoinDelta.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
+package ch.uzh.ifi.hase.soprafs22.lobby;
+import ch.uzh.ifi.hase.soprafs22.game.player.IPlayer;
+
+public class PlayerJoinDelta {
```
Comment (paolovic) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/PlayerJoinDelta.java:4]: I'd suggest "LobbyDelta" similar to GameDelta

Comment (apple00juice) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/lobby/PlayerJoinDelta.java:4]: Sure, I'll change it

**Review by paolovic — APPROVED** [review]

**Pull request merged**